src/worker/activities/local_activities.rs

```diff
@@ -0,0 +1,170 @@
+    /// Constrains number of currently executing local activities
+    semaphore: Semaphore,
+    /// Sink for new activity execution requests
+    act_req_tx: UnboundedSender<NewLocalAct>,
+
+    dat: Mutex<LAMData>,
+}
+
+struct LAMData {
+    /// Activities that need to be executed by lang
+    act_req_rx: UnboundedReceiver<NewLocalAct>,
+    /// Activities that have been issued to lang but not yet completed
+    outstanding_activity_tasks: HashMap<TaskToken, InFlightLocalActInfo>,
+    next_tt_num: u32,
+}
+
+impl LocalActivityManager {
+    pub(crate) fn new(max_concurrent: usize) -> Self {
+        let (act_req_tx, act_req_rx) = unbounded_channel();
+        Self {
+            semaphore: Semaphore::new(max_concurrent),
+            act_req_tx,
+            dat: Mutex::new(LAMData {
+                act_req_rx,
+                outstanding_activity_tasks: Default::default(),
+                next_tt_num: 0,
+            }),
+        }
+    }
+
+    pub(crate) fn enqueue(&self, acts: impl IntoIterator<Item = NewLocalAct> + Debug) {
+        debug!("Queuing local activities: {:?}", &acts);
+        for act in acts {
+            self.act_req_tx
+                .send(act)
+                .expect("Receive half of LA request channel cannot be dropped");
+        }
+    }
+
+    pub(crate) async fn next_pending(&self) -> Option<ActivityTask> {
+        // Wait for a permit to take a task
+        let permit = self.semaphore.acquire().await.expect("is never closed");
+        let mut dat = self.dat.lock().await;
+        // It is important that there are no await points after receiving from the channel, as
+        // it would mean dropping this future would cause us to drop the activity request.
+        if let Some(new_la) = dat.act_req_rx.recv().await {
+            let sa = new_la.schedule_cmd;
+
+            dat.next_tt_num += 1;
+            let tt = TaskToken::new_local_activity_token(dat.next_tt_num.to_le_bytes());
+            dat.outstanding_activity_tasks.insert(
+                tt.clone(),
+                InFlightLocalActInfo {
+                    seq: sa.seq,
+                    workflow_execution: new_la.workflow_exec_info.clone(),
+                },
+            );
+
+            // Forget the permit. Permits are removed until a completion.
+            permit.forget();
+
+            Some(ActivityTask {
+                task_token: tt.0,
+                activity_id: sa.activity_id,
+                variant: Some(activity_task::Variant::Start(Start {
+                    workflow_namespace: sa.namespace,
+                    workflow_type: new_la.workflow_type,
+                    workflow_execution: Some(new_la.workflow_exec_info),
+                    activity_type: sa.activity_type,
+                    header_fields: sa.header_fields,
+                    input: sa.arguments,
+                    heartbeat_details: vec![],
+                    // TODO: Get these times somehow
+                    scheduled_time: None,
```

Comment:
Does this mean schedule-to-close timeouts will not currently work?

Reply:
As of this PR yeah. All this time handling stuff I will do in one of the subsequent ones